core/arch/arm/kernel/early_ta.c

```diff
@@ -0,0 +1,249 @@
+{
+	const struct early_ta *ta;
+
+	for_each_early_ta(ta)
+		if (!memcmp(&ta->uuid, uuid, sizeof(*uuid)))
+			return ta;
+
+	return NULL;
+}
+
+static void *zalloc(void *opaque __unused, unsigned int items,
+		    unsigned int size)
+{
+	return malloc(items * size);
+}
+
+static void zfree(void *opaque __unused, void *address)
+{
+	free(address);
+}
+
+static bool decompression_init(z_stream *strm,
+			       const struct early_ta *ta)
+{
+	int st;
+
+	strm->next_in = ta->ta;
+	strm->avail_in = ta->size;
+	strm->zalloc = zalloc;
+	strm->zfree = zfree;
+	st = inflateInit(strm);
+	if (st != Z_OK) {
+		EMSG("Decompression initialization error (%d)", st);
+		return false;
+	}
+
+	return true;
+}
+
+static TEE_Result early_ta_open(const TEE_UUID *uuid,
+				struct user_ta_store_handle **h)
+{
+	struct user_ta_store_handle *handle;
+	const struct early_ta *ta;
+	bool st;
+
+	ta = find_early_ta(uuid);
+	if (!ta)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	handle = calloc(1, sizeof(*handle));
+	if (!handle)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (ta->uncompressed_size) {
+		/* TA is compressed */
+		st = decompression_init(&handle->strm, ta);
+		if (!st) {
+			free(handle);
+			return TEE_ERROR_BAD_FORMAT;
+		}
+	}
+	handle->early_ta = ta;
+	*h = handle;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result early_ta_get_size(const struct user_ta_store_handle *h,
+				    size_t *size)
+{
+	const struct early_ta *ta = h->early_ta;
+
+	if (ta->uncompressed_size)
+		*size = ta->uncompressed_size;
+	else
+		*size = ta->size;
+	*size = h->early_ta->size;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result read_uncompressed(struct user_ta_store_handle *h, void *data,
+				    size_t len)
+{
+	uint8_t *src = (uint8_t *)h->early_ta->ta + h->offs;
+
+	if (h->offs + len > h->early_ta->size)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if (data)
+		memcpy(data, src, len);
+	h->offs += len;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result read_compressed(struct user_ta_store_handle *h, void *data,
+				  size_t len)
+{
+	z_stream *strm = &h->strm;
+	size_t total = 0;
+	uint8_t *tmpbuf = NULL;
+	TEE_Result ret;
+	size_t out;
+	int st;
+
+	if (data) {
+		strm->next_out = data;
+		strm->avail_out = len;
+	} else {
+		/*
+		 * inflate() does not support a NULL strm->next_out. So, to
+		 * discard data, we have to allocate a temporary buffer. 1K
+		 * seems reasonable.
+		 */
+		strm->avail_out = MIN(len, 1024U);
+		tmpbuf = malloc(strm->avail_out);
+		if (!tmpbuf) {
+			EMSG("Out of memory");
+			return TEE_ERROR_OUT_OF_MEMORY;
+		}
+		strm->next_out = tmpbuf;
+	}
+	/*
+	 * Loop until we get as many bytes as requested, or an error occurs.
+	 * inflate() returns:
+	 * - Z_OK when progress was made, but neither the end of the input
+	 *   stream nor the end of the output buffer were met
+	 * - Z_BUF_ERROR when there is still input to process but the output
+	 *   buffer is full
+	 * - Z_STREAM_END when the end of the intput stream was reached
```

Comment:
Reading again the current comment, I found it clear. Maybe move `Z_STREAM_END` description next to `Z_OK`. It would look less like an error case (as right after `Z_BUF_ERROR`);

Minor: add a terminal dots to each bullet.